Updates per #291 #292 #294

qchaupds opened the pull request:
Allows Line Feed (LF) character to terminate a record in table format file and resolves associated file even if schematron is not available in a label.

Closes NASA-PDS/validate#291, NASA-PDS/validate#292 and NASA-PDS/validate#294


This pull request resolves three issues:

1.  If the schematron is not available, the label parsing and validation still take place,
2. The line feed character is now allowed to end a record,
3. The comparisons of floating points is corrected when the min and max values are the same as the record content.

Tested in DEV:

To test NASA-PDS/validate#291

% validate -R pds4.bundle  -r report_github291_bundle_valid.json   -s json -t src/test/resources/github291/valid/bundle_kaguya_derived.xml
% validate -R pds4.bundle  -r report_github291_bundle_invalid.json -s json -t src/test/resources/github291/invalid/bundle_kaguya_derived.xml


To test NASA-PDS/validate#292:

Test a valid label, one with Line Feed (LF) and one with Carriage Return/Line Feed (CR/LF) record terminator.

% validate --skip-context-validation -R pds4.label  -r report_github292_label_valid_lf.json     -s json -t src/test/resources/github292/valid/minimal_test_product_lf.xml
% validate --skip-context-validation -R pds4.label  -r report_github292_label_valid_crlf.json   -s json -t src/test/resources/github292/valid/minimal_test_product_crlf.xml

Test a invalid label, one with LF and one with CR/LF record terminator.  The error will be because the line length specified (999) is too long (60 actual length).

% validate --skip-context-validation -R pds4.label  -r report_github292_label_invalid_lf.json   -s json -t src/test/resources/github292/invalid/minimal_test_product_lf.xml
% validate --skip-context-validation -R pds4.label  -r report_github292_label_invalid_crlf.json -s json -t src/test/resources/github292/invalid/minimal_test_product_crlf.xml

To test NASA-PDS/validate#294:

% validate --skip-context-validation -R pds4.label -r report_github294_bundle_invalid.json -s json -t src/test/resources/github294/invalid/minmax-error.xml
% validate --skip-context-validation -R pds4.label -r report_github294_bundle_valid.json   -s json -t src/test/resources/github294/valid/minmax-error.xml

Commit:
Handle bad schematron definition in label, allow LF character as record delimiter and enhance floating-point comparisons

1. Add test artifacts github291, github292 and github294 to src/test/resources directory
2. Add comments for future changes to determining if a label is a bundle or a collection based on filename to LabelValidator.java
3. Modify conversion from text to Double instead of Number to match types of 'mininum' and 'maximum' variables in FieldValueValidator.java
4. Force parse and validate a label even if schematron is not available in LabelValidationRule.java
5. Add check if a record ends with a line feed in TableDataContentValidationRule.java
6. Correct logic to check for optional field 'field_format' in TableFieldDefinitionRule.java
7. Add more enum types and add checking for "validate#87" test to build catalog file in StepDefs.java
8. Add tests for 3 new issues #291, #292 and #294 to file validate.feature

Refs #291 https://github.com/NASA-PDS/validate/issues/291 When validating a product with a bad schematron definition, bundle validation also fails indicating the associated product does not exist
     #292 https://github.com/NASA-PDS/validate/issues/292 CCB-264: Make the Line Feed (LF) character an allowed record delimiter
     #294 https://github.com/NASA-PDS/validate/issues/294 Content validation incorrectly reports error for floating-point values out of specified min/max range

Commit:
Remove commented out code.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:
```diff
@@ -37,6 +37,7 @@
 
 import gov.nasa.pds.tools.label.ExceptionType;
 import gov.nasa.pds.tools.util.FileSizesUtil;
+import gov.nasa.pds.tools.util.FileReferencedMapList;
```
@qchaupds looks like you are missing this class in your commit?

Commit:
Add missing class FileReferencedMapList.java

jordanpadams committed:
Update test data with 1.16.0.0 schemas

1. the labels must refer to the correct record_delimiter in order to validate properly
2. this is only valid as of pds v1.16.0.0 (1G00), so we need to include those schemas in the test data for now.

refs #292

jordanpadams reviewed (changes requested):
@qchaupds I thought I sent this in our email discussions, but can't track it down.

#292 requires we validate against an unreleased version of the IM (1.16.0.0 aka 1G00). i've added those to the repo.

additionally, validate needs to validate according to the line delimiter specified in `//record_delimiter`. for example, for a `Table_Character`:
```
    <Table_Character>
      <name>Test Instrument Table Data</name>
      <offset unit="byte">0</offset>
      <records>4</records>
      <description>This table is a Character table</description>
      <record_delimiter>Carriage-Return Line-Feed</record_delimiter>
```
vs.
```
    <Table_Character>
      <name>Test Instrument Table Data</name>
      <offset unit="byte">0</offset>
      <records>4</records>
      <description>This table is a Character table</description>
      <record_delimiter>Line-Feed</record_delimiter>
```
This will need to apply to all tables.

i've updated the test data to reflect the new IM version, and the Line-Feed for those test products. however, I think the software will need to be updated to key off of that `//record_delimiter` attribute value.

here is how you can run the software to specify schema/schematrons via command-line:

```
validate-1.25.0-SNAPSHOT/bin/validate -t src/test/resources/github292/valid/minimal_test_product_crlf.xml \
    --schema  src/test/resources/github292/PDS4_PDS_1G00.xsd \
    --schematron src/test/resources/github292/PDS4_PDS_1G00.sch \
    --skip-context-validation
```

Commit:
Response to code change request from pull request: Modify test code to get cucumber testing to work

1. Add test artifacts PDS4_PDS_1G00.sch and PDS4_PDS_1G00.xsd to src/test/resources/github292/pds/v1 directory to match convention
2. Add writing of catalog files to StepDefs.java
3. Rearrange many tests in file validate.feature and corrected some
4. Correct 'data_type' tag value to reflect actual format of ASCII_Date_Time_YMD_UTC

Refs #291 https://github.com/NASA-PDS/validate/issues/291 When validating a product with a bad schematron definition, bundle validation also fails indicating the associated product does not exist
     #292 https://github.com/NASA-PDS/validate/issues/292 CCB-264: Make the Line Feed (LF) character an allowed record delimiter
     #294 https://github.com/NASA-PDS/validate/issues/294 Content validation incorrectly reports error for floating-point values out of specified min/max range

Commit:
Remove comment.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -426,14 +447,23 @@ public void validateTableDataContents() throws MalformedURLException,
               progressCounter();
+              lineNumber += 1;
               
               if (!line.endsWith("\r\n")) {
```
@qchaupds per #292, I think the logic can be much simpler here. most importantly, per this ticket, the check for line ending must be keyed off of the value in the `record_delimiter` attribute. here is some pseudocode for a simple check:
```
// Very similar to the original logic in this method, 
// except we first read the value of the <record_delimiter> attribute
if record_delimiter.value() == "Carriage-Return Line-Feed" AND !line.endsWith("\r\n"):
    throw ERROR
// Next, we need to do an else if because CRLF ends in a new line so we need to do some additional checks
else if record_delimiter.value() == "Line-Feed":
    if (line.endsWith("\r\n") OR !line.endsWith("\n"):
        throw an ERROR
```

Per the sample data, here is an example  of where you can find `record_delimiter`:
```
  <File_Area_Observational>
    <File>
      <file_name>minimal_test_product_crlf.tab</file_name>
      <file_size unit="byte">240</file_size>
      <md5_checksum>5bc560bb3554b8d705ec36e3905726f5</md5_checksum>
    </File>

    <Table_Character>
      <name>Test Instrument Table Data</name>
      <offset unit="byte">0</offset>
      <records>4</records>
      <description>This table is a Character table</description>
      <record_delimiter>Carriage-Return Line-Feed</record_delimiter>
```

jordanpadams replied:
additionally, I don't think we care about the ASCII value. you should just be able to just check if the string ends with `\r\n` or just `\n`. Let me know if that makes sense? or we can setup some time tomorrow to discuss

Commit:
Second response to code change request from pull request: Check for line ending to match what was provided in the label

1. Add new class to get tag record_delimiter in TabulatedUtil.java
2. Add new enumerated type MISSING_LF in ProblemType.java
2. Modify logic to check for line ending based on what was provided in label in TableDataContentValidationRule.java

Refs #291 https://github.com/NASA-PDS/validate/issues/291 When validating a product with a bad schematron definition, bundle validation also fails indicating the associated product does not exist
     #292 https://github.com/NASA-PDS/validate/issues/292 CCB-264: Make the Line Feed (LF) character an allowed record delimiter
     #294 https://github.com/NASA-PDS/validate/issues/294 Content validation incorrectly reports error for floating-point values out of specified min/max range

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/TabulatedUtil.java:
```diff
@@ -0,0 +1,63 @@
+      List<TinyNodeImpl> xmlModels = new ArrayList<TinyNodeImpl>();
+      try {
+          SAXSource saxSource = new SAXSource(Utility.openConnection(getTarget()));
+          DocumentInfo docInfo = LabelParser.parse(saxSource); // Parses a label.
```
@qchaupds apologies for poking at this more but I've come across a few more issues

---

- [x] **issue 1**:

we have to read in the label to DOM for other aspects of content validation as well. for instance:

1. here: https://github.com/NASA-PDS/validate/blob/master/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java#L125
2. and here: https://github.com/NASA-PDS/validate/blob/master/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java#L125 https://github.com/NASA-PDS/validate/blob/master/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java#L167

I'm not sure if the read on L125 can be set as a global or not, but if not, at minimum we should refactor the code with L167 so we read in the XML object and can use it for your check + the remainder of the code.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -451,28 +436,51 @@ public void validateTableDataContents() throws MalformedURLException,
-                    dataFile,
-                    tableIndex,
-                    reader.getCurrentRow());              
+                  // If cannot find a record delimiter, check for the default carriage return line feed.
```
@qchaupds if record_delimiter is null or not in ("Carriage-Return Line-Feed", "Line-Feed") this is an invalid label. schematron validation will throw an error, but we can just add another error here so we fail gracefully on content validation as well.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/TabulatedUtil.java:
```diff
@@ -0,0 +1,63 @@
+          SAXSource saxSource = new SAXSource(Utility.openConnection(getTarget()));
+          DocumentInfo docInfo = LabelParser.parse(saxSource); // Parses a label.
+          XMLExtractor extractor = new XMLExtractor(docInfo);
+          xmlModels = extractor.getNodesFromDoc("Product_Observational/File_Area_Observational/Table_Character");
```
- [x] 1. if we could use the `Xpaths` class to define these XPaths as variables that would great. see https://github.com/NASA-PDS/validate/blob/master/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java#L128 for an example of usage. hard coding in these classes can be really hard to track down in the future.

- [x] 2. This does not just apply to `Table_Character`, this applies to all places where `record_delimiter` exists in a product. if possible, I would just generalize this to `//record_delimiter`. if that is not possible, then we will need to find all the objects this applies to in the [PDS4 IM Spec](https://pds.nasa.gov/datastandards/documents/im/v1/index_1G00.html)

qchaupds replied:
As it turns out, another class contains a function to fetch the delimiter already so there's no need to re-parse the label file.

Commit:
Make the fetching of field 'record_delimiter' more general to all labels.

Commit:
Use getRecordDelimiter() from TableCharacter class to get the record_delimiter tag.

jordanpadams committed:
updated test data for effective integration testing

* updated all labels to remove optional checksum / file size to make it easier to update data products for testing purposes
* updated LF.tab files to no longer include ASCII Line-Feed character. they should simply end in a newline `\n`. required an update to the record_length as well since we are removing the CR character.
* updated invalid test cases to be more in scope of checking CRLF vs LF
  * CRLF test case now looks at an LF.tab and doesn't find the CRLF, and throws appropriate error
  * LF test case now looks at a CRLF.tab with the assumed record length for just LF, which throws appropriate several errors regarding missing LF and record length

jordanpadams committed:
Additional test data for other product types

* added test data for tabled_delimited, inventory, and stream_text
* re-organized test data

Notes:
* stream_text is not currently included in content validation. tests will need to be updated once #229 is completed
* per the SCR inventory products may not be included in this update. awaiting a check there
* this is not the comprehensive list of all possible Table_* this could apply to, since there are several Table_* classes that inherit from Table_Delimited / Table_Character

jordanpadams committed:
fix bug in test data

jordanpadams committed:
remove duplicate schemas

jordanpadams reviewed (changes requested):
## Review
We are getting closer. The test you have look good. I just updated the data a bit for the invalid tests to be more CRLF v LF-centric to make sure they are testing against one another.
 
I then added a whole bundle other test data for other types of tables that are included with this update to the IM. More specifically:
•	Table_Delimited
•	Stream_Text.
•	Inventory
 
For Table_Delimited, this should be similar implementation to Table_Character, you will just need to handle that use case. I have hoping that since Inventory extends Table_Delimited, once  you make the update to work for the Table_Delimited test cases, it will work with Inventory. If that is the case, this will make our lives a lot easier for all the other possible extensions of these tables. If that is not the case and we should discuss our options.
 
For Stream_Text, I wouldn’t worry too much about this right now. We do not test the content of these products right now, and that is not the end of the world. We have another ticket in the backlog we will get to to address this.

Additionally, the DDWG just noted that if a file specifies a LF in the label, but it ends in a CRLF, we should catch this and throw an WARNING since it is unlikely this was intended. A new test label, `src/test/resources/github292/table_character/invalid/minimal_test_product_lf_2.xml` was created to test this. Right now, validate says this is valid. However, even though the records are 60 characters long (valid) and end in a `\n` (valid), technically there is a CRLF at the end of the line, so we should catch this and throw a WARNING.

## Work Items
Here is a checklist of items that need to be done:

- [x] update tests to use new directory structure

See checklist tagged in the code for additional items.

jordanpadams committed:
test label for testing CRLF is invalid when LF is specified

for this test data, the records are 60 characters long (valid) and end in a `\n` (valid) but technically there is a CRLF, so we should catch this and throw an error.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -426,23 +431,55 @@ public void validateTableDataContents() throws MalformedURLException,
+                        "Record does not end in carriage-return line feed.",
+                        dataFile, tableIndex, reader.getCurrentRow());
+                    manuallyParseRecord = true;
+                  } else if (recordDelimiter.equalsIgnoreCase("Line-Feed") && !line.endsWith("\n")) {
```
@qchaupds per my review comments, we need to improve this logic.

- [x] update logic for handling CRLF line endings when LF is specified in the label

```
if recordDelimiter == Line-Feed:
   if !line.endsWith("\n"): THROW ERROR   // this is your current error thrown
   if line.endsWith("\r\n"): THROW ERROR   // this is a new error. label says LF, but line ends in CRLF
```

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -340,6 +344,7 @@ public void validateTableDataContents() throws MalformedURLException,
         } else {
           LOG.debug("table instanceof TableCharacter: else");
           TableCharacter tc = (TableCharacter) table;
```
@qchaupds per comments in review, we need to add additional logic in the code about for handling record delimiters for `Table_Delimited`:

- [x] update code and add tests for table_delimited

right now we only handle this for `Table_Character` per your updates below.

Commit:
Third response to code change request from pull request: Add tests for table_delimited and inventory files

1. Add new logic to process table_delimited and inventory files in TableDataContentValidationRule.java
2. Add new enumerated types in StepDefs.java
3. Fix failing table_delimited test and add new inventory files test in validate.feature
4. Fix line feed end character (add Control-L) in inventory/collection_eetable_inventory_LF.csv
4. Fix line feed end character (add Control-L) in stream_text/aareadme_LF.txt
4. Fix line feed end character (add Control-L) in table_delimited/kgrs_calibrated_spectra_per1_LF.csv

Refs #291 https://github.com/NASA-PDS/validate/issues/291 When validating a product with a bad schematron definition, bundle validation also fails indicating the associated product does not exist
     #292 https://github.com/NASA-PDS/validate/issues/292 CCB-264: Make the Line Feed (LF) character an allowed record delimiter
     #294 https://github.com/NASA-PDS/validate/issues/294 Content validation incorrectly reports error for floating-point values out of specified min/max range

Commit:
Correct the end character to be just the line feed character.

Commit:
Forth  response to code change request from pull request: Add tests for table_characters and fix end charactes.

1. Add 4 table_characters to validate.feature
2. Remove extract Control-L character in inventory/collection_eetable_inventory_LF.csv
3. Remove extract Control-L character in stream_text/aareadme_LF.txt
4. Remove extract Control-L character in table_character/invalid/minimal_test_product_lf.tab
5. Correct bug with wrong values of record_delimiter value and record_length in table_character/valid/minimal_test_product_crlf.xml
6. Remove extract Control-L character in table_delimited/kgrs_calibrated_spectra_per1_LF.csv

Refs #291 https://github.com/NASA-PDS/validate/issues/291 When validating a product with a bad schematron definition, bundle validation also fails indicating the associated product does not exist
     #292 https://github.com/NASA-PDS/validate/issues/292 CCB-264: Make the Line Feed (LF) character an allowed record delimiter
     #294 https://github.com/NASA-PDS/validate/issues/294 Content validation incorrectly reports error for floating-point values out of specified min/max range

jordanpadams committed:
Update errors messages
